src/server/page_server.cpp

```diff
@@ -913,22 +978,92 @@ page_server::disconnect_all_tran_servers ()
   er_log_debug (ARG_FILE_LINE, "disconnect_all_tran_server: All connections have been disconnected.\n");
 }
 
+void
+page_server::disconnect_all_follower_page_servers ()
+{
+  /*
+   * TODO it should request to disconnect to all followers and wait until it's done.
+   * It will be addressed when follower has a way to re-start the catch-up with another followee.
+   * For now, this waits until just all the catch-up jobs in progress are done.
+   */
+  constexpr auto millis_20 = std::chrono::milliseconds { 20 };
+
+  er_log_debug (ARG_FILE_LINE, "Wait until all follower connections are disconnected.\n");
+
+  auto ulock_conn_vec = std::unique_lock { m_follower_conn_vec_mutex };
+  auto lockg_disc = std::lock_guard { m_follower_disc_mutex };
+
+  while (!m_follower_conn_vec_cv.wait_for (ulock_conn_vec, millis_20, [this]
+  {
+    return m_follower_conn_vec.empty ();
+    }));
+
+  assert (m_follower_conn_vec.empty());
+
+  if (m_follower_disc_future.valid ())
+    {
+      m_follower_disc_future.get (); // wait until it's done if there is an on-going disconnection.
+    }
+  er_log_debug (ARG_FILE_LINE, "All follower connections have been disconnected.\n");
+}
+
 void
 page_server::disconnect_followee_page_server (const bool with_disc_msg)
 {
-  auto lockg = std::lock_guard <std::mutex> (m_followee_conn_mutex);
+  auto lockg = std::lock_guard { m_followee_conn_mutex };
   if (m_followee_conn != nullptr)
     {
+      auto channel_id = m_followee_conn->get_channel_id ();
       if (with_disc_msg)
 	{
-	  //TODO send the disconnect msg
+	  m_followee_conn->push_request (follower_to_followee_request::SEND_DISCONNECT_MSG, std::string ());
 	}
       m_followee_conn.reset (nullptr);
 
-      er_log_debug (ARG_FILE_LINE, "The followee page server has been disconnected.\n");
+      er_log_debug (ARG_FILE_LINE, "The followee page server has been disconnected. channel id: %s\n", channel_id.c_str ());
     }
 }
 
+void
+page_server::disconnect_follower_page_server_async (const follower_connection_handler *conn)
+{
+  follower_connection_handler_uptr_t disconnecting_conn_uptr;
+  {
+    auto lockg_conn_vec = std::lock_guard { m_follower_conn_vec_mutex };
+
+    auto it = std::find_if (m_follower_conn_vec.begin (), m_follower_conn_vec.end (),
+			    [&conn] (auto & conn_uptr)
+    {
+      return conn_uptr.get () == conn;
+    });
+
+    if (it == m_follower_conn_vec.cend ())
+      {
+	// It's already been disconnected or in progress by another thread.
+	return;
+      }
+    disconnecting_conn_uptr = std::move (*it);
```

Comment:
nit
cppreference says:
```suggestion
    disconnecting_conn_uptr.reset ((*it).release ());
```

Reply:
> 1) Move assignment operator. Transfers ownership from r to *this as if by calling reset(r.release()) followed by assigning [get_deleter()](https://en.cppreference.com/w/cpp/memory/unique_ptr/get_deleter) from [std::forward](http://en.cppreference.com/w/cpp/utility/forward)<Deleter>(r.get_deleter()).
from https://en.cppreference.com/w/cpp/memory/unique_ptr/operator%3D

What I've found out is that the `move` is for this kind of job, including releasing ownership, transfer, and even moving deleter. And I saw some arguments that prefer move over release, and I relate to it. Could you give me the cppference that you mentioned?